### PR TITLE
[ArrayManager] TST: arithmetic test

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
xref https://github.com/pandas-dev/pandas/issues/39146/

### CODE REVIEW

File: pandas/_testing/__init__.py [hunk 1]
```diff
@@ -207,8 +207,10 @@ def box_expected(expected, box_cls, transpose=True):
         if transpose:
             # for vector operations, we need a DataFrame to be a single-row,
             #  not a single-column, in order to operate against non-DataFrame
-            #  vectors of the same length.
+            #  vectors of the same length. But convert to two rows to avoid
+            #  single-row special cases in datetime arithmetic
             expected = expected.T
+            expected = pd.concat([expected] * 2, ignore_index=True)
```
Comment (jorisvandenbossche) [inline pandas/_testing/__init__.py:213]: @jbrockmendel by making this a (2, x) DataFrame instead of a (1, x) shaped DataFrame, it avoids some special cases in the tests where we now often have "if box is DataFrame -> don't raise a PerformanceWarning" (which is quite confusing, because it has nothing to do with it being a DataFrame, but rather with the column arrays being length 1)

**Review by jreback — APPROVED** [review]

File: pandas/_testing/__init__.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_testing/__init__.py:213]: seems Mostly Harmless i guess

File: pandas/tests/arithmetic/test_datetime64.py [hunk 2]
```diff
@@ -318,40 +318,40 @@ def test_dt64arr_timestamp_equality(self, box_with_array):
             box_with_array if box_with_array not in [pd.Index, pd.array] else np.ndarray
         )
 
-        ser = Series([Timestamp("2000-01-29 01:59:00"), "NaT"])
+        ser = Series([Timestamp("2000-01-29 01:59:00"), Timestamp("2000-01-30"), "NaT"])
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_datetime64.py:321]: why is this necessary?

File: pandas/tests/arithmetic/test_timedelta64.py [hunk 3]
```diff
@@ -2071,6 +2077,11 @@ def test_td64arr_div_numeric_array(self, box_with_array, vector, any_real_dtype)
             result = tdser / vector.astype(object)
             if box_with_array is pd.DataFrame:
                 expected = [tdser.iloc[0, n] / vector[n] for n in range(len(vector))]
+                if using_array_manager:
+                    # https://github.com/pandas-dev/pandas/issues/39750
+                    # third column with all-NaT as result doesn't get preserved
+                    # as timedelta64 dtype
+                    result[2] = pd.array(["NaT", "NaT"], dtype="timedelta64[ns]")
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_timedelta64.py:2084]: my intuition is we should edit expected, not result.  is there a reason you did it this way?

File: pandas/tests/arithmetic/test_datetime64.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_datetime64.py:321]: Because I made the dataframe (2, N) shaped, a series with 2 elements would give a (2, 2) dataframe (and a column would be length 2), which would no longer need alignment, changing the test (eg it no longer would warn).

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:2084]: Just because there was already a `if box_with_array is pd.DataFrame` branch here, but can also add one below after the final `expected` is created and edit that one.

Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py]: updated

File: pandas/tests/arithmetic/test_timedelta64.py [hunk 4]
```diff
@@ -2075,6 +2081,13 @@ def test_td64arr_div_numeric_array(self, box_with_array, vector, any_real_dtype)
                 expected = [tdser[n] / vector[n] for n in range(len(tdser))]
             expected = pd.Index(expected)  # do dtype inference
             expected = tm.box_expected(expected, xbox)
+
+            if using_array_manager and box_with_array is pd.DataFrame:
+                # https://github.com/pandas-dev/pandas/issues/39750
+                # third column with all-NaT as result doesn't get preserved
+                # as timedelta64 dtype
+                expected[2] = Series([pd.NaT, pd.NaT], dtype=object)
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_timedelta64.py:2089]: this is sub-optimal.  can you add a TODO to try to retain the correct dtype

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_timedelta64.py]: thanks

**Comment by jbrockmendel** [comment]
one request https://github.com/pandas-dev/pandas/pull/39753#discussion_r575583968, otherwise LGTM

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:2089]: I don't fully understand your comment. Of course it's sub-optimal, as it's a bug (but an existing one, and I opened an issue about it)

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_timedelta64.py:2089]: > but an existing one

now im confused; we only get here if `using_array_manager`

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_timedelta64.py:2089]: i.e. if the expected isnt actually what we want, better to xfail

Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:2089]: > now im confused; we only get here if using_array_manager

Yes, I "uncovered" the bug by adding array manager tests, but the bug itself is not related to ArrayManager, it's a bug in TimedeltaArray (and you can run into it without ArrayManager as well, just not covered by any test)

But indeed, an xfail might be more appropriate since I am now asserting the buggy behaviour. The problem is that I would then no longer test the other parts of this test (the other columns) which is actually working fine. 
So I would prefer to keep it this way (but will update the comment to make it more clear this is buggy)

Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:2091]: As mentioned above, and already included in the comment as well, I already opened an issue (it's https://github.com/pandas-dev/pandas/issues/39750)

File: pandas/tests/arithmetic/test_datetime64.py [hunk 5]
```diff
@@ -1276,7 +1273,7 @@ def test_dt64arr_add_sub_relativedelta_offsets(self, box_with_array):
             ]
         )
         vec = tm.box_expected(vec, box_with_array)
-        vec_items = vec.squeeze() if box_with_array is pd.DataFrame else vec
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_datetime64.py:1279]: This is not about ArrayManager not supporting `squeeze`, this is just because I changed the dataframe from shape (1, 3) to (2, 3), so to get the first row as a Series, `squeeze` now no longer is suited in general (this is not related to ArrayManager vs BlockManager)

File: pandas/tests/arithmetic/test_numeric.py [hunk 6]
```diff
@@ -532,13 +532,15 @@ def test_df_div_zero_series_does_not_commute(self):
...
+    def test_df_mod_zero_df(self, using_array_manager):
         # GH#3590, modulo as ints
         df = pd.DataFrame({"first": [3, 4, 5, 8], "second": [0, 0, 0, 3]})
 
         # this is technically wrong, as the integer portion is coerced to float
-        # ###
-        first = Series([0, 0, 0, 0], dtype="float64")
+        first = Series([0, 0, 0, 0])
+        if not using_array_manager:
+            # BlockManager doesn't preserve dtype per column if possible
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_numeric.py:542]: I am right now adding `# TODO(ArrayManager)` for things I still need to fix, and I can add a similar ``# INFO(ArrayManager)`` for cases like the above where behaviour changed but this is actually OK and not TODO to fix. Does that sound good?

File: pandas/tests/arithmetic/test_timedelta64.py [hunk 7]
```diff
@@ -1761,6 +1763,8 @@ def test_td64arr_floordiv_td64arr_with_nat(self, box_with_array):
 
         expected = np.array([1.0, 1.0, np.nan], dtype=np.float64)
         expected = tm.box_expected(expected, xbox)
+        if box is DataFrame and using_array_manager:
+            expected[[0, 1]] = expected[[0, 1]].astype("int64")
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_timedelta64.py:1767]: why is expected different and is one behavior more desirable than the other?

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_timedelta64.py:2091]: > and you can run into it without ArrayManager as well, just not covered by any test

OK.  Can you add a non-ArrayManager test that will hit it?

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:1767]: ArrayManager performs the op column-wise, and thus can preserve dtypes (floordiv gives ints). Will add a comment about it in the line of https://github.com/pandas-dev/pandas/pull/39753/files#r576350947

Comment (jbrockmendel) [reply pandas/tests/arithmetic/test_timedelta64.py:1767]: thanks

File: pandas/tests/arithmetic/test_timedelta64.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/tests/arithmetic/test_timedelta64.py:2094]: I added additional examples that don't involve ArrayManager to the issue
